misc/nsis/install.nsh

```diff
@@ -0,0 +1,555 @@
+
+  !insertmacro MULTIUSER_GetCurrentUserString $0
+  CreateShortCut "$DESKTOP\${PRODUCT_NAME}$0.lnk" "$INSTDIR\${PROGEXE}"
+SectionEnd
+
+Section "Start Menu Icon" SectionStartMenuIcon
+  SectionIn 1 2
+
+  !insertmacro MULTIUSER_GetCurrentUserString $0
+  CreateShortCut "$STARTMENU\${PRODUCT_NAME}$0.lnk" "$INSTDIR\${PROGEXE}"
+SectionEnd
+
+SectionGroupEnd
+
+Section "-Write Install Info" ; hidden section, write install info as the final step
+  !insertmacro MULTIUSER_RegistryAddInstallSizeInfo
+  !insertmacro MULTIUSER_GetCurrentUserString $0
+  WriteRegStr SHCTX "${MULTIUSER_INSTALLMODE_UNINSTALL_REGISTRY_KEY_PATH}$0" "HelpLink" "${HELP_LINK}"
+  WriteRegStr SHCTX "${MULTIUSER_INSTALLMODE_UNINSTALL_REGISTRY_KEY_PATH}$0" "URLInfoAbout" "${URL_ABOUT}"
+  WriteRegStr SHCTX "${MULTIUSER_INSTALLMODE_UNINSTALL_REGISTRY_KEY_PATH}$0" "URLUpdateInfo" "${URL_UPDATE}"
+  WriteRegStr SHCTX "${MULTIUSER_INSTALLMODE_UNINSTALL_REGISTRY_KEY_PATH}$0" "Comments" "${COMMENTS}"
+  WriteRegStr SHCTX "${MULTIUSER_INSTALLMODE_UNINSTALL_REGISTRY_KEY_PATH}$0" "Contact" "${CONTACT}"
+
+  ; Add InstallDate String
+  System::Call /NOUNLOAD '*(&i2,&i2,&i2,&i2,&i2,&i2,&i2,&i2) i .r9'
+  System::Call /NOUNLOAD 'kernel32::GetLocalTime(i)i(r9)'
+  System::Call /NOUNLOAD '*$9(&i2,&i2,&i2,&i2,&i2,&i2,&i2,&i2)i(.r1,.r2,.r3,.r4,.r5,.r6,.r7,)'
+  System::Free $9
+  IntCmp $2 9 0 0 +2
+  StrCpy $2 '0$2'
+  IntCmp $4 9 0 0 +2
+  StrCpy $4 '0$4'
+  WriteRegStr SHCTX "${MULTIUSER_INSTALLMODE_UNINSTALL_REGISTRY_KEY_PATH}$0" "InstallDate" "$1$2$4"
+
+  ; Refresh Shell Icons
+  System::Call "shell32::SHChangeNotify(i 0x08000000, i 0, i 0, i 0)"
+SectionEnd
+
+; Modern install component descriptions
+!insertmacro MUI_FUNCTION_DESCRIPTION_BEGIN
+  !insertmacro MUI_DESCRIPTION_TEXT ${SectionCoreFiles} \
+    "Core files required to run ${PRODUCT_NAME}."
+  !insertmacro MUI_DESCRIPTION_TEXT ${SectionGroupIntegration} \
+    "Integrate ${PRODUCT_NAME} with the Operating System."
+  !insertmacro MUI_DESCRIPTION_TEXT ${SectionWindowsRegister} \
+    "Register protocols and file extensions with ${PRODUCT_NAME}."
+  !insertmacro MUI_DESCRIPTION_TEXT ${SectionDefaultBrowser} \
+    "Set ${PRODUCT_NAME} as the default Web browser."
+  !insertmacro MUI_DESCRIPTION_TEXT ${SectionGroupShortcuts} \
+    "Create shortcut icons to run ${PRODUCT_NAME}."
+  !insertmacro MUI_DESCRIPTION_TEXT ${SectionDesktopIcon} \
+    "Create ${PRODUCT_NAME} icon on the Desktop."
+  !insertmacro MUI_DESCRIPTION_TEXT ${SectionStartMenuIcon} \
+    "Create ${PRODUCT_NAME} icon in the Start Menu."
+!insertmacro MUI_FUNCTION_DESCRIPTION_END
+
+; Callbacks
+Function .onInit
+  StrCpy $KeepReg 1
+  !insertmacro CheckPlatform ${PLATFORM}
+  !insertmacro CheckMinWinVer ${MIN_WIN_VER}
+  ${ifnot} ${UAC_IsInnerInstance}
+    !insertmacro CheckSingleInstance "Setup" "Global" "${SETUP_MUTEX}"
+    !insertmacro CheckSingleInstance "Application" "Local" "${APP_MUTEX}"
+  ${endif}
+
+  ; Detect existing setup from previous installers
```

Comment:
Thats odd, i never had a hang when testing upgrades/reinstalls. The fact that `uninst.exe` (the name of the previous NSISuninstaller) still existed is also strange (unless it was `unistall.exe` - the new name). This file is deleted by the script if it is found in `$INSTDIR` as it's usually leftover: because Windows can't delete open files, the NSIS uninstaller in order to be able to delete itself, it first copies itself to `$TMP`, then runs again from there (passing also `$INSTDIR`) and exits the original process. This has the drawback of "confusing" `ExecWait`, so when the `_?` parameter is passed the uninstaller executes (without copying) from `$INSTDIR` and then is left behind.

What happened in this case, is that for some reason the `uninstall.exe` got lost but the registry still points to it. If you try to remove qutebrowser from `Add/Remove Programs`, Windows will inform about the invalid registry and prompt to fix the problem (or you can manualy delete `HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Uninstall\qutebrowser`).

It would be good idea for the installer to detect and fix that problem though, do you agree?

Reply:
Also note that this part of the code is for detecting installations of the previous NSIS installer.
I made a change in 7fb0c85 so when the new uninstaller is missing the upgrade won't fail.